### PR TITLE
Add table azure_data_factory. Closes #156

### LINKED ISSUE #156: Add table azure_data_factory
**References**
https://github.com/Azure/azure-sdk-for-go/tree/main/services/datafactory/mgmt/2018-06-01/datafactory

### PR DESCRIPTION
**Pull request by rajeshbal65**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_data_factory []

PRETEST: tests/azure_data_factory

TEST: tests/azure_data_factory
Running terraform
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 4s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest69476]
azurerm_data_factory.named_test_resource: Creating...
azurerm_data_factory.named_test_resource: Still creating... [10s elapsed]
azurerm_data_factory.named_test_resource: Creation complete after 15s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest69476/providers/Microsoft.DataFactory/factories/turbottest69476]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Warning: Version constraints inside provider configuration blocks are deprecated

  on variables.tf line 22, in provider "azurerm":
  22:   version         = "=1.36.0"

Terraform 0.13 and earlier allowed provider version constraints inside the
provider configuration block, but that is now deprecated and will be removed
in a future version of Terraform. To silence this warning, move the provider
version constraint into the required_providers block.


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

resource_aka = "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest69476/providers/Microsoft.DataFactory/factories/turbottest69476"
resource_aka_lower = "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest69476/providers/microsoft.datafactory/factories/turbottest69476"
resource_id = "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest69476/providers/Microsoft.DataFactory/factories/turbottest69476"
resource_name = "turbottest69476"
subscription_id = "d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8"

Running SQL query: test-get-query.sql
[
  {
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest69476/providers/Microsoft.DataFactory/factories/turbottest69476",
    "name": "turbottest69476",
    "region": "eastus",
    "resource_group": "turbottest69476",
    "subscription_id": "d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8",
    "tags": {
      "name": "turbottest69476"
    },
    "type": "Microsoft.DataFactory/factories"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest69476/providers/Microsoft.DataFactory/factories/turbottest69476",
    "name": "turbottest69476",
    "type": "Microsoft.DataFactory/factories"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest69476/providers/Microsoft.DataFactory/factories/turbottest69476",
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest69476/providers/microsoft.datafactory/factories/turbottest69476"
    ],
    "name": "turbottest69476",
    "tags": {
      "name": "turbottest69476"
    },
    "title": "turbottest69476"
  }
]
✔ PASSED

POSTTEST: tests/azure_data_factory

TEARDOWN: tests/azure_data_factory

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  name,
  id,
  type,
  provisioning_state,
  etag
from
  azure_data_factory;

+-------------+------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+--------------------+----------------------------------------+
| name        | id                                                                                                                                 | type                            | provisioning_state | etag                                   |
+-------------+------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+--------------------+----------------------------------------+
| test-rajesh | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbot_rg/providers/Microsoft.DataFactory/factories/test-rajesh | Microsoft.DataFactory/factories | Succeeded          | "90017c20-0000-0100-0000-60dbff910000" |
+-------------+------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+--------------------+----------------------------------------+
```

### List system assigned identity type Factories

```sql
select
  name,
  id,
  type,
  identity ->> 'type' as identity_type
from
  azure_data_factory
where
  identity ->> 'type' = 'SystemAssigned';

+-------------+------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+
| name        | id                                                                                                                                 | type                            | identity_type  |
+-------------+------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+
| test-rajesh | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbot_rg/providers/Microsoft.DataFactory/factories/test-rajesh | Microsoft.DataFactory/factories | SystemAssigned |
+-------------+------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+----------------+
```

### List factories with plubic network access is allowed

```sql
select
  name,
  id,
  type,
  public_network_access
from
  azure_data_factory
where
  public_network_access = 'Enabled';

+------+----+------+-----------------------+
| name | id | type | public_network_access |
+------+----+------+-----------------------+
+------+----+------+-----------------------+
```
```
</details>

### CODE REVIEW

File: azure/table_azure_data_factory.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/datafactory/mgmt/2018-06-01/datafactory"
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+)
+
+//// TABLE DEFINITION ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:13]:
```suggestion
//// TABLE DEFINITION
```

File: azure/table_azure_data_factory.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+			},
+			{
+				Name:        "provisioning_state",
+				Description: "Factory provisioning state, example Succeeded.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("FactoryProperties.ProvisioningState"),
+			},
+			{
+				Name:        "create_time",
+				Description: "Time the factory was created in ISO8601 format.",
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:58]:
```suggestion
				Description: "Specifies the time, the factory was created.",
```

File: azure/table_azure_data_factory.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+			{
+				Name:        "version",
+				Description: "Version of the factory.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("FactoryProperties.Version"),
+			},
+			{
+				Name:        "public_network_access",
+				Description: "Whether or not public network access is allowed for the data factory.",
+				Type:        proto.ColumnType_STRING,
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:71]: Is the type correct?
Seems like, the type should be `boolean`

File: azure/table_azure_data_factory.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+				Transform:   transform.FromField("FactoryProperties.RepoConfiguration"),
+			},
+			{
+				Name:        "global_parameters",
+				Description: "List of parameters for factory.",
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("FactoryProperties.GlobalParameters"),
+			},
+
+			// Standard columns
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:98]:
```suggestion
			// Steampipe standard columns
```

File: azure/table_azure_data_factory.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+				Description: ColumnDescriptionTags,
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+			{
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:116]:
```suggestion
			
			// Azure standard columns
			{
```

File: azure/table_azure_data_factory.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+				Name:        "subscription_id",
+				Description: ColumnDescriptionSubscription,
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ID").Transform(idToSubscriptionID),
+			},
+		},
+	}
+}
+
+//// LIST FUNCTIONS ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:138]:
```suggestion
//// LIST FUNCTION
```

File: azure/table_azure_data_factory.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		for _, factory := range result.Values() {
+			d.StreamListItem(ctx, factory)
+		}
+		result.NextWithContext(context.Background())
+		pagesLeft = result.NotDone()
+	}
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:166]:
```suggestion
//// HYDRATE FUNCTIONS
```

File: azure/table_azure_data_factory.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+func getFactory(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getFactory")
+
+	name := d.KeyColumnQuals["name"].GetStringValue()
+	resourceGroup := d.KeyColumnQuals["resource_group"].GetStringValue()
+
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:173]: Add check for empty input

File: docs/tables/azure_data_factory.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  type,
+  identity ->> 'type' as identity_type
+from
+  azure_data_factory
+where
+  identity ->> 'type' = 'SystemAssigned';
+```
+
+
+### List factories with plubic network access is allowed
```
Comment (Subhajit97) [inline docs/tables/azure_data_factory.md:36]:
```suggestion
### List factories with public network access enabled
```

File: docs/tables/azure_data_factory.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  id,
+  type,
+  provisioning_state,
+  etag
+from
+  azure_data_factory;
+```
+
+
+### List system assigned identity type Factories
```
Comment (Subhajit97) [inline docs/tables/azure_data_factory.md:21]:
```suggestion
### List system assigned identity type factories
```

File: azure/table_azure_data_factory.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+//// TABLE DEFINITION
+
+func tableAzureDataFactory(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "azure_data_factory",
+		Description: "Azure Data Factory",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.AllColumns([]string{"name", "resource_group"}),
+			Hydrate:           getFactory,
+			ShouldIgnoreError: isNotFoundError([]string{"ResourceNotFound", "ResourceGroupNotFound", "404"}),
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:22]: Validate error codes
Do we require both `ResourceNotFound` and `404`?

File: azure/table_azure_data_factory.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+				Name:        "version",
+				Description: "Version of the factory.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("FactoryProperties.Version"),
+			},
+			{
+				Name:        "public_network_access",
+				Description: "Whether or not public network access is allowed for the data factory.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("FactoryProperties.PublicNetworkAccess"),
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:72]:
```suggestion
				Transform:   transform.FromField("FactoryProperties.PublicNetworkAccess").Transform(transform.ToString),
```

File: azure/table_azure_data_factory.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+	op, err := factoryClient.Get(ctx, resourceGroup, name, "*")
+	if err != nil {
+		return nil, err
+	}
+
+	// In some cases resource does not give any notFound error
+	// instead of notFound error, it returns empty data
+	if op.ID != nil {
+		return op, nil
+	}
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:198]: Is this check required?

File: docs/tables/azure_data_factory.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  type,
+  identity ->> 'type' as identity_type
+from
+  azure_data_factory
+where
+  identity ->> 'type' = 'SystemAssigned';
+```
+
+
+### List factories with publicbic network access enabled
```
Comment (Subhajit97) [inline docs/tables/azure_data_factory.md:36]:
```suggestion
### List factories with public network access enabled
```

File: azure-test/tests/azure_data_factory/test-get-expected.json [hunk 15]
```diff
@@ -0,0 +1,13 @@
+[
+  {
+    "id": "{{ output.resource_id.value }}",
+    "name": "{{resourceName}}",
+    "region": "eastus",
```
Comment (Subhajit97) [inline azure-test/tests/azure_data_factory/test-get-expected.json:5]: don't use hardcoded value

**Review by Subhajit97 — APPROVED** [review]

**Review by rajlearner17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**